components/msg/background_hang_monitor.rs

```diff
@@ -2,50 +2,244 @@
+pub type MonitoredThreadId = mach::mach_types::thread_act_t;
+
+#[cfg(target_os = "windows")]
+pub type MonitoredThreadId = usize; // TODO: use winapi
+
+#[cfg(any(target_os = "android", target_os = "linux"))]
+pub type MonitoredThreadId = libc::pthread_t;
+
+type Address = *const libc::uint8_t;
+
+/// The registers used for stack unwinding
+struct Registers {
+    /// Instruction pointer.
+    pub instruction_ptr: Address,
+    /// Stack pointer.
+    pub stack_ptr: Address,
+    /// Frame pointer.
+    pub frame_ptr: Address,
+}
+
+struct NativeStack {
+    instruction_ptrs: [*mut std::ffi::c_void; MAX_NATIVE_FRAMES],
+    stack_ptrs: [*mut std::ffi::c_void; MAX_NATIVE_FRAMES],
+    count: usize
+}
+
+impl NativeStack {
+    fn new() -> Self {
+        NativeStack {
+            instruction_ptrs: [ptr::null_mut(); MAX_NATIVE_FRAMES],
+            stack_ptrs: [ptr::null_mut(); MAX_NATIVE_FRAMES],
+            count: 0
+        }
+    }
+
+    fn process_register(&mut self, instruction_ptr: *mut std::ffi::c_void, stack_ptr: *mut std::ffi::c_void) {
+        self.instruction_ptrs[self.count] = instruction_ptr;
+        self.stack_ptrs[self.count] = stack_ptr;
+        self.count = self.count + 1;
+    }
+}
+
+#[cfg(target_os = "macos")]
+fn check_kern_return(kret: mach::kern_return::kern_return_t, success: &mut bool) {
+    if kret != mach::kern_return::KERN_SUCCESS {
+        warn!("Kern return error: {:?}", kret);
+        *success = false;
+    }
+}
+
+#[cfg(target_os = "macos")]
+#[allow(unsafe_code)]
+unsafe fn suspend_thread(thread_id: MonitoredThreadId, success: &mut bool) {
+    check_kern_return(mach::thread_act::thread_suspend(thread_id), success);
 }
 
+#[cfg(target_os = "windows")]
 #[allow(unsafe_code)]
-#[cfg(not(windows))]
-unsafe fn get_backtrace_from_monitored_component(monitored: &MonitoredComponent) -> Option<Backtrace> {
-    libc::pthread_kill(monitored.thread_id, libc::SIGPROF);
+unsafe fn suspend_thread(thread_id: MonitoredThreadId, success: &mut bool) {
+    // TODO: use winapi::um::processthreadsapi::SuspendThread
+    *success = false;
+}
+
+#[cfg(any(target_os = "android", target_os = "linux"))]
+#[allow(unsafe_code)]
+unsafe fn suspend_thread(thread_id: MonitoredThreadId, success: &mut bool) {
+    match libc::pthread_kill(thread_id, libc::SIGPROF) {
+        0 => {},
+        _ => {
+            *success = false;
+        },
+    };
+}
+
+#[cfg(target_os = "macos")]
+#[allow(unsafe_code)]
+unsafe fn get_registers(thread_id: MonitoredThreadId, regs: &mut Registers, success: &mut bool) {
+    let state = mach::structs::x86_thread_state64_t::new();
+    let state_count = mach::structs::x86_thread_state64_t::count();
+    let kret = mach::thread_act::thread_get_state(thread_id,
+                                                  mach::thread_status::x86_THREAD_STATE64,
+                                                  mem::transmute(&state),
+                                                  mem::transmute(&state_count));
```

Comment:
These are transmuting immutable pointers to mutable ones; the values would be declared with `mut` and we should pass &mut to the first transmute and have no need of the second transmute.